### PR TITLE
#647: Added HINCRBY Command

### LINKED ISSUE #647: Add support for command `HINCRBY`
Add support for the `HINCRBY` command in DiceDB similar to the [`HINCRBY` command in Redis](https://redis.io/docs/latest/commands/command-list/). Please refer to the following commit in Redis to understand the implementation specifics - [source](https://github.com/redis/redis/tree/f60370ce28b946c1146dcea77c9c399d39601aaa).

Write unit and integration tests for the command referring to the tests written in the [Redis codebase 7.2.5](https://github.com/redis/redis/tree/f60370ce28b946c1146dcea77c9c399d39601aaa). For integration tests, you can refer to the `tests` folder. Note: they have used `TCL` for the test suite, and we need to port that to our way of writing integration tests using the relevant helper methods. Please refer to our [tests directory](https://github.com/DiceDB/dice/tree/master/tests).

For the command, benchmark the code and measure the time taken and memory allocs using [`benchmem`](https://blog.logrocket.com/benchmarking-golang-improve-function-performance/) and try to keep them to the bare minimum.

### PR DESCRIPTION
**Pull request by ankkyprasad**
Closes #647

### CODE REVIEW

File: internal/eval/eval.go
Comment (apoorvyadav1111) [inline internal/eval/eval.go:2760]: Hi, upon check the redis behavior for handling this command, I think we can order the validation to ensure we are consistent with the output when there are multiple issues with the command. The order in redis is:
1. Check arity
2. Check value is integer
3. Get or Create object and validate type
4. Check field is integer
5. perform operation
I have following suggestion, if you would consider it:
```suggestion

	field := args[1]
	increment, err := strconv.ParseInt(args[2], 10, 64)
	if err != nil {
		return diceerrors.NewErrWithFormattedMessage(diceerrors.IntOrOutOfRangeErr)
	}
	key := args[0]
	obj := store.Get(key)
	var hashmap HashMap

	if obj != nil {
		if err := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeHashMap, object.ObjEncodingHashMap); err != nil {
			return diceerrors.NewErrWithMessage(diceerrors.WrongTypeErr)
		}
		hashmap = obj.Value.(HashMap)
	}

	if hashmap == nil {
		hashmap = make(HashMap)
	}


```

File: internal/eval/hmap.go
Comment (apoorvyadav1111) [inline internal/eval/hmap.go:100]: Another suggestion here would be to check overflow after incrementing. Please check the following scenario:
image

File: internal/eval/commands.go
Comment (apoorvyadav1111) [inline internal/eval/commands.go:602]: nit:
```suggestion
		Arity:    3,
```

File: integration_tests/commands/hincrby_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/hincrby_test.go:14]: Would it be okay to add a test case when incrementing with two big yet 64 bit integers would overflow. This will help Dice to check every scenario.
nit: name field in test cases help future contributors to understand about the test cases.

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]

File: integration_tests/commands/hincrby_test.go
Comment (AshwinKul28) [reply integration_tests/commands/hincrby_test.go:14]: Yes please add all edge cases which you can think of. Don't hesitate, the more the number, the more we are safe.

File: integration_tests/commands/hincrby_test.go
Comment (AshwinKul28) [inline integration_tests/commands/hincrby_test.go:37]: Try to add a case for -ve value. It should decrement it.

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]

**Comment by AshwinKul28** [comment]
LGTM @ankkyprasad Thanks a lot for this command support. Left few comments, have a look.

File: internal/eval/eval.go
Comment (ankkyprasad) [reply internal/eval/eval.go:2760]: Why does the ordering of the validations matter here? All the validations are independent of each other. So. the ordering shouldn't effect the operation being performed.

File: internal/eval/hmap.go
Comment (ankkyprasad) [reply internal/eval/hmap.go:100]: My Bad. I missed this edgecase. Thank you for pointing it out.

File: internal/eval/eval.go
Comment (apoorvyadav1111) [reply internal/eval/eval.go:2760]: Hi, the ordering of the validations performed  will decide what error message you will be returning when multiple validations fail. For instance, lets consider the following run in redis:
image

when I run the command `hincrby a` and `hincrby a v`. It returns Arity error. 

If I run the command `hincrby a v a`, it returns integer error. In your case, since the object does not exist, it will create and then check and then return integer error.

However, If I create an object which is of different type using cmd `set a b` and then run again, as per your logic, it will return type error, but as you see in the output it returns integer error.

For reference please check code for redis [here](https://github.com/redis/redis/blob/5f28bd96dbe67e5d924d5bbab3ada5618ec52490/src/t_hash.c#L2187).

File: integration_tests/commands/async/hincrby_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/async/hincrby_test.go:50]: Hi, Changes look good, Can you please add on last test case with following scenario:
"HINCRBY should give integer error when trying to increment a key which is not a hash value with a value which is not integer"

Thanks,
Apoorv

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]

**Comment by AshwinKul28** [comment]
@ankkyprasad Thanks a lot for the quick changes. It's good to have one more test case which @apoorvyadav1111 suggested above. Please take care of that. Otherwise, the changes look great. Also, one more thing - of you can come up with the comment above eval function about what it does in a generalized way in 3-4 lines that would be great. 

Please rebase your branch with the master to resolve merge conflicts. thanks

**Comment by AshwinKul28** [comment]
@ankkyprasad Looks good. Please check the test cases.

File: integration_tests/commands/async/hincrby_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/async/hincrby_test.go:72]: Hi, Thanks for adding this test case. Please the expected output in redis.
image
We need to return an integer error in this case.

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]

**Comment by apoorvyadav1111** [comment]
Hi @ankkyprasad, Upon checking the check logs, I realized we are not cleaning up after running test cases, which is why the hset tests are failing. For tests to not affect further tests, you can run `del` commands as clean-up or `flushdb` as the last test case. This will ensure the next test will have no conflicting keys.

I think with these changes, we have a solid implementation for this command. Please let me know if you need more clarification on cleanup. 

Apoorv

**Comment by ankkyprasad** [comment]
> Hi @ankkyprasad, Upon checking the test suite logs, I realized we are not cleaning up after running test cases, which is why the hset tests are failing. For tests to not affect further tests, you can run `del` commands as clean-up or `flushdb` as the last test case. This will ensure the next test will have no conflicting keys.
> 
> I think with these changes, we have a solid implementation for this command. Please let me know if you need more clarification on cleanup.
> 
> Apoorv

Hey Apoorv, 

Thank you so much. I was trying to debug from last night. You saved my day.

**Comment by ankkyprasad** [comment]
Let's review and merge it off before another merge conflict comes! 😭 
@apoorvyadav1111 @AshwinKul28

**Comment by JyotinderSingh** [comment]
Just as a side note, it's always better to run cleanup commands as a part of defer instead of adding them as a separate test case. I've made the changes for you in this PR.

**Review by apoorvyadav1111 — APPROVED** [review]
Looks good to me. Thanks for addressing all of my comments. I do not have permissions to merge it though. However, one of the maintainers will be doing it soon.

**Comment by ankkyprasad** [comment]
> Just as a side note, it's always better to run cleanup commands as a part of defer instead of adding them as a separate test case. I've made the changes for you in this PR.

Understood! 

Will take care of this from next time

**Pull request merged**
